formats/json/jsMain/src/kotlinx/serialization/json/internal/DynamicDecoders.kt

```diff
@@ -79,7 +79,7 @@ private open class DynamicInput(
     }
 
     override fun decodeTaggedEnum(tag: String, enumDescriptor: SerialDescriptor): Int =
-        enumDescriptor.getElementIndexOrThrow(getByTag(tag) as String)
+        enumDescriptor.getJsonNameIndexOrThrow(json, getByTag(tag) as String)
```

Comment:
Could you please also add a test for a dynamic decoder?

Reply:
It turns out DynamicDecoder does not support JsonNames at all, so I'll leave it for future PR

Reply:
And it also does not support coercing